components/tidb_query/src/codec/convert.rs

```diff
@@ -228,47 +230,72 @@ impl ToInt for u64 {
     fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
         let upper_bound = integer_unsigned_upper_bound(tp);
         if *self > upper_bound {
-            ctx.handle_overflow_err(overflow!(self, upper_bound))?;
+            ctx.handle_overflow_err(overflow(self, tp))?;
             return Ok(upper_bound);
         }
         Ok(*self)
     }
 }
 
 impl ToInt for f64 {
+    /// This function is ported from TiDB's types.ConvertFloatToInt,
+    /// which checks whether the number overflows the signed lower and upper boundaries of `tp`
+    ///
+    /// # Notes
+    ///
+    /// It handles overflows using `ctx` so that the caller would not handle it anymore.
     fn to_int(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<i64> {
+        #![allow(clippy::float_cmp)]
         let val = (*self).round();
         let lower_bound = integer_signed_lower_bound(tp);
         if val < lower_bound as f64 {
-            ctx.handle_overflow_err(overflow!(val, lower_bound))?;
+            ctx.handle_overflow_err(overflow(val, tp))?;
             return Ok(lower_bound);
         }
 
         let upper_bound = integer_signed_upper_bound(tp);
-        if val > upper_bound as f64 {
-            ctx.handle_overflow_err(overflow!(val, upper_bound))?;
-            return Ok(upper_bound);
+        let ub_f64 = upper_bound as f64;
+        // according to https://github.com/pingcap/tidb/pull/5247
+        if val >= ub_f64 {
+            if val == ub_f64 {
+                return Ok(upper_bound);
+            } else {
+                ctx.handle_overflow_err(overflow(val, tp))?;
+                return Ok(upper_bound);
+            }
         }
         Ok(val as i64)
     }
 
+    /// This function is ported from TiDB's types.ConvertFloatToUint,
+    /// which checks whether the number overflows the unsigned upper boundaries of `tp`
+    ///
+    /// # Notes
+    ///
+    /// It handles overflows using `ctx` so that the caller would not handle it anymore.
+    #[allow(clippy::float_cmp)]
     fn to_uint(&self, ctx: &mut EvalContext, tp: FieldTypeTp) -> Result<u64> {
         let val = (*self).round();
         if val < 0f64 {
-            ctx.handle_overflow_err(overflow!(val, 0))?;
+            ctx.handle_overflow_err(overflow(val, tp))?;
             if ctx.should_clip_to_zero() {
                 return Ok(0);
             } else {
+                // recall that, `f64 as u64` is different from `f64 as i64 as u64`
                 return Ok(val as i64 as u64);
             }
         }
-
         let upper_bound = integer_unsigned_upper_bound(tp);
         if val > upper_bound as f64 {
-            ctx.handle_overflow_err(overflow!(val, upper_bound))?;
-            return Ok(upper_bound);
+            ctx.handle_overflow_err(overflow(val, tp))?;
+            Ok(upper_bound)
+        } else if val == upper_bound as f64 {
```

Comment:
Seems we can simplify this code:
```rust
if val > upper_bound as f64 {
    ctx.handle_overflow_err(overflow(val, tp))?;
    return Ok(upper_bound);	            Ok(upper_bound)
} else {
    Ok(val as u64)
}
```
See: https://gist.github.com/4c3bcdc9ec36c60bd5920f6be138918b